syntax = "proto3";

message NormalizerConfig{
  bool lower = 1;
  string unicode_format = 2;
  repeated CustomRule custom_rules = 3;
}

message CustomRule {
  string pattern = 1;
  string replacement = 2;
}

message UnigramConfig {
  uint32 vocab_seed = 1;
  uint32 max_sub_len = 2;
  uint32 em_iters = 3;
  float n_per = 4;
  float epsilon = 5;
  float alpha = 6;
  float unk_penalty = 7;
}

message UnigramProb {
  string token = 1;
  float prob = 2;
}

message VocabEntry {
  string key = 1;
  uint32 value = 2;
}

message UNK {
  string token = 1;
  uint32 value = 2;
}

message PAD {
  string token = 1;
  uint32 value = 2;
}

message SOS {
  string token = 1;
  uint32 value = 2;
}

message EOS {
  string token = 1;
  uint32 value = 2;
}

message DefaultSpecialToken {
  PAD pad = 1;
  UNK unk = 2;
  SOS sos = 3;
  EOS eos = 4;
}

message CustomTokenEntry {
  string sp_token = 1;
  uint32 value = 2;
}

message ModelProto {
  repeated VocabEntry vocab = 1;
  DefaultSpecialToken special_tokens = 2;
  repeated CustomTokenEntry custom_tokens = 3; 
  NormalizerConfig normalizer = 4;
  string model_type = 5;
  uint32 vocab_size = 6;
  UnigramConfig unigram = 7;
  repeated UnigramProb unigram_prob = 8;
}